packages/predicate/src/predicate.ts

```diff
@@ -111,18 +112,31 @@ export class Predicate<ARGS extends InputValue[]> extends Account implements Abs
    * @returns The Predicate instance with updated predicate data.
    */
   setData<T extends ARGS>(...args: T) {
+    this.predicateArgs = args;
+
+    return this;
+  }
+
+  #getPredicateData(txLength: number): Uint8Array {
+    if (!this.predicateArgs.length) {
+      return new Uint8Array();
+    }
+
     const mainFn = this.interface?.functions.main;
     const paddedCode = new ByteArrayCoder(this.bytes.length).encode(this.bytes);
 
     const VM_TX_MEMORY = calculateVmTxMemory({
       maxInputs: this.provider.getChain().consensusParameters.maxInputs.toNumber(),
     });
-    const SEMI_OFFSET = VM_TX_MEMORY + INPUT_COIN_FIXED_SIZE + WORD_SIZE + paddedCode.byteLength;
-
-    this.getPredicateData = (txLength: number) =>
-      mainFn?.encodeArguments(args, SEMI_OFFSET + txLength) || new Uint8Array();
-
-    return this;
+    const OFFSET =
+      VM_TX_MEMORY +
+      SCRIPT_FIXED_SIZE +
+      INPUT_COIN_FIXED_SIZE +
+      WORD_SIZE +
+      paddedCode.byteLength +
+      txLength;
```

Comment:
Following previous suggestions.

```suggestion
      (policiesLength * WORD_SIZE);
```